pkg/kubelet/server/server.go

```diff
@@ -688,6 +706,53 @@ func (s *Server) getSpec(request *restful.Request, response *restful.Response) {
 	response.WriteEntity(info)
 }
 
+func createPod(host string, resources []v1.ResourceRequirements) *v1.Pod {
+	containers := make([]v1.Container, 0)
+	for _, resource := range resources {
+		containers = append(containers, v1.Container { Resources: resource, })
+	}
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "podToTest",
+		},
+		Spec: v1.PodSpec{
+			NodeName: host,
+			Containers: containers,
+
+		},
+		Status: v1.PodStatus{
+			Conditions: []v1.PodCondition{
+				{
+					Type:   v1.PodScheduled,
+					Status: v1.ConditionTrue,
+				},
+			},
+		},
+	}
+}
+
+func (s *Server) canAssignPod(request *restful.Request, response *restful.Response) {
+	//only one container just for the sake of measurements
+	resources := make([]v1.ResourceRequirements, 0)
```

Comment:
I think we should not make dummy container spec or pod spec here.

The reason is, Topology Manager allocates resources while calculating topology hint.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/topology_manager.go#L232
The logic of `Admit` function is :
- for containers in Pod Spec(iteration)
  - getTopologyHints
  - merge Hints then determine admission
  - allocate //it allocate actual resources for pod

So if we using dummy Pod here, hint provider will allocate actual resources for dummy Pod.
And the allocated resources for dummy Pod will be released by garbage collector(reconciler) of each hint provider(CPU Manager and Device Manager).
(In reconciler it compares activePods with Pods that the hint provider allocates resources, if reconciler found stale pod, it frees the resource allocation of stale pod, see here for [reconciler](https://github.com/kubernetes/kubernetes/blob/f9532e46632f487363d8e1b499f92c854b905189/pkg/kubelet/cm/cpumanager/cpu_manager.go#L306) for cpu manager, see [here](https://github.com/kubernetes/kubernetes/blob/f9532e46632f487363d8e1b499f92c854b905189/pkg/kubelet/cm/devicemanager/manager.go#L618) for reconciler for device manager)

So I would say `canAssignPod` would call `CheckPodAdditions` with Pod spec in http reuqest.

Reply:
@bg-chun thanks, this answers my question above re: resource deallocation. However this seems to suggest that with the current code the only thing to do would be to start the pod in the first node that could accomodate the pod. There is no real option to try to fit a pod but not run it unless we do further refactoring in the topology manager, and I'm not sure how invasive that could be.

Reply:
@fromanirh 
Maybe the below sequence diagram is a little bit helpful to understand how topology manager works.
https://docs.google.com/presentation/d/1ea4SlauSMrVnXCOtVaFm0LpG5-Y8afEH3O3wmXqyXjs/edit?usp=sharing

Reply:
I excluded Allocate from RESTfull call path. Reservation would be done by following way: NodeInfo.RequestedResource from scheduler would be passed to kubelet each time we make a request. CPU/DeviceManager should take into account this value.

Reply:
Does Reservation mean a excluding resources from v1.node object?

Reply:
No, Reservation here means taking into account resources for the pod which was already assigned to the node (Bind), but not yet started. I saw your comment in https://docs.google.com/presentation/d/1SR6XSIFsHkiWTws66LABpTiZaRwYk8IoRts4HQWE8Bc/edit#slide=id.g74d9a6aad0_0_109. I hurried, skipAllocate works only for pod level resources.